### PR TITLE
Associate Camel K schema when modeline is available

### PR DESCRIPTION
**Pull request by apupier**
completionWithCamelKModelineOnly

### CODE REVIEW

**Comment by apupier** [comment]
this is failing due to https://github.com/redhat-developer/vscode-yaml/issues/323

**Comment by apupier** [comment]
job is green on CI, github is not updatign the status: https://travis-ci.org/github/camel-tooling/vscode-camelk/builds/719861172

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ All notable changes to the "vscode-camelk" extension will be documented in this
 
 ## 0.0.17
 
-- TBD
+- Associate yaml Camel K schema when there is the Camel K modeline
```
Comment (lhein) [inline CHANGELOG.md:7]: should we write it with capital letters? Like Yaml or YAML ?

File: src/CamelKSchemaManager.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import fetch from 'cross-fetch';
+
+const CAMELK_MODELINE_PREFIX = '# camel-k:';
+const CAMELK_SCHEMA_ID = 'camelk';
+const CAMELK_SCHEMA_URI_PREFIX = CAMELK_SCHEMA_ID + '://schema/';
+
+let camelkSchemaCache: string;
+
+export async function registerCamelKSchemaProvider(mainOutputChannel: vscode.OutputChannel) {
+	const yamlExtension = vscode.extensions.getExtension('redhat.vscode-yaml');
```
Comment (lhein) [inline src/CamelKSchemaManager.ts:29]: can you specify types please?

File: src/CamelKSchemaManager.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	await retrieveSchemaAsCache(mainOutputChannel);
+	if(yamlExtension?.isActive) {
+		const yamlExtensionAPI = yamlExtension?.exports;
```
Comment (lhein) [inline src/CamelKSchemaManager.ts:32]: can you specify types please?

File: src/CamelKSchemaManager.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		yamlExtensionAPI.registerContributor(CAMELK_SCHEMA_ID, requestYamlSchemaUriCallback, requestYamlSchemaContentCallback);
+	} else {
+		mainOutputChannel.appendLine('Yaml extension is not active. There won\'t be automatic completion/validation for Camel K files.');
+	}
+}
+
+async function retrieveSchemaAsCache(mainOutputChannel: vscode.OutputChannel) {
+	try {
+		const res = await fetch(getCamelKSchemaUrl());
```
Comment (lhein) [inline src/CamelKSchemaManager.ts:41]: can you specify types please?

File: src/CamelKSchemaManager.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		if(res.status >= 400) {
+			mainOutputChannel.appendLine('Cannot retrieve Camel K schema. '+ res.statusText);
+			return Promise.resolve(undefined);
```
Comment (lhein) [inline src/CamelKSchemaManager.ts:44]: you can simple return `undefined` imo.

File: src/CamelKSchemaManager.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	await retrieveSchemaAsCache(mainOutputChannel);
+	if(yamlExtension?.isActive) {
+		const yamlExtensionAPI = yamlExtension?.exports;
+		yamlExtensionAPI.registerContributor(CAMELK_SCHEMA_ID, requestYamlSchemaUriCallback, requestYamlSchemaContentCallback);
+	} else {
+		mainOutputChannel.appendLine('Yaml extension is not active. There won\'t be automatic completion/validation for Camel K files.');
+	}
+}
+
+async function retrieveSchemaAsCache(mainOutputChannel: vscode.OutputChannel) {
```
Comment (lhein) [inline src/CamelKSchemaManager.ts:39]: can you specify the return type of the function?

File: src/CamelKSchemaManager.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+			camelkSchemaCache = await res.text();
+			return camelkSchemaCache;
+		}
+	} catch(err) {
+		mainOutputChannel.appendLine('Cannot retrieve Camel K schema. '+ err);
+	}
+}
+
+function requestYamlSchemaUriCallback(resource: string): string | undefined {
+	const textDocument = vscode.workspace.textDocuments.find(document => {
```
Comment (lhein) [inline src/CamelKSchemaManager.ts:55]: can you specify types please?

File: src/CamelKSchemaManager.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+		return document.uri.toString() === resource;
+	});
+	if (textDocument
+		&& resource.endsWith('.yaml')
+		&& textDocument.getText().startsWith(CAMELK_MODELINE_PREFIX)) {
+			return CAMELK_SCHEMA_URI_PREFIX + resource;
+		}
+	return undefined;
```
Comment (lhein) [inline src/CamelKSchemaManager.ts:63]: I think that is the default if you don't specify it, but I could be wrong.

File: src/CamelKSchemaManager.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+function requestYamlSchemaContentCallback(uri: string): string | undefined {
+	const parsedUri = vscode.Uri.parse(uri);
```
Comment (lhein) [inline src/CamelKSchemaManager.ts:67]: can you specify types please?

File: src/CamelKSchemaManager.ts (hunk 8 above)
Comment (apupier) [reply src/CamelKSchemaManager.ts:63]: if you are not sure, it means that it is easier to understand when specifying it

Comment (lhein) [reply src/CamelKSchemaManager.ts:63]: yeah, was just thinking loud. I am fine with the current way you handle it.

**Comment by apupier** [comment]
> What's the use case for this? User creates a new, blank yaml and wants to create a camel k file?

Camel K developers doesn't want to impose *.camelk.yaml file name pattern.

> Not arguing that it's not helpful, but wondering where the user is coming at this from.

coming at request from Luca

> If they create a new yaml Camel K file through the Camel K extension, this is done for them already, right?

no, when creating from Command Palette, it is reusing the kamel init command and then it is creating with <theNameprovidedByUser>.yaml and not <thenameprovidedByUser>.camelk.yaml

**Review by lhein — APPROVED** [review]

**Pull request merged**